Automattic/wp-calypso
# POC: Site Assembler supports adding customized fonts and colors

#### Proposed Changes

**Demo**

https://user-images.githubusercontent.com/13596067/211771917-6f794f3e-aab5-4917-a3a3-18cdbd6f4ee3.mov

**Concepts**

* Introduce the `@automattic/global-styles` package to re-implement `GlobalStylesProvider` on the calypso side. We will update the global styles when the user chooses any color palette or font pairing.
* We treat any color palette or font pairing as a variation to provide powerful control of the global styles. That is, a font pairing option can define not only the font family but also the font size, font weight, etc.
* We get the customized styles via `useGlobalStylesOutput` and inject the generated CSS via `EditorStyles`. Additionally, the output of the `useGlobalStylesOutput` is the pure CSS array, so we can also inject the inline CSS manually to display the customized styles.
* After we finish the site assembler step, we can send the final user config to the global styles endpoint to update the user styles right after we activate the blank-canvas-3.

**Issues need to resolve**

* We need to upgrade `@wordpress/edit-site` to `4.9.0` above to support changing the font family of the heading via `elements.heading` as it requires the version of `@wordpress/blocks` is `11.11.0` above
* The font pairing option doesn't display the correct font family because there is no way to load the font family from the client side. Now the font families are loaded from the injected inline style such as `wp-webfonts-jetpack-google-fonts` from the server side
* The performance is slow as any style change will trigger **EVERY** pattern to re-render including patterns shown on the large preview. So, we're not able to render all patterns at once, and we need to render the pattern when it shows. But if the user selects too many patterns, the issue will happen again because of the large preview.
* For now the global styles endpoint only supports getting the styles of the active theme. See [here](https://github.com/WordPress/wordpress-develop/blob/d2b385b7b82c092d564cae6d07e1e456c014e558/src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php#L549)
* I'm not sure why I cannot read the user config of global styles on my site. See [here](https://github.com/WordPress/wordpress-develop/blob/d2b385b7b82c092d564cae6d07e1e456c014e558/src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php#L135)

#### Testing Instructions



* Activate the BC3 theme first as the global styles endpoint only supports the active theme
* Go to /setup?siteSlug=<your_site>
* Click the "Continue" button until you land on the Design Picker step
* Scroll to the bottom and select the Blank Canvas CTA
* When you're in the Pattern Assembler step
  * Select a pattern first
  * Choose a color palette to see the color/background/... of the large preview is changed
  * Choose a font pairing to see the font of the large preview is changed

#### Pre-merge Checklist





Related to https://github.com/Automattic/wp-calypso/issues/71038
